### PR TITLE
Additive Fore→Aft Cross-Attention in AftSRF: per-node wake context

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The AftSRF correction head (#2104, merged) independently refines each aft-foil surface node using its Transolver hidden state plus base prediction. It has no direct, spatially-resolved access to **what the fore foil looks like** — yet the aft-foil pressure is governed by the fore-foil wake impinging on it.

PR #2202 (closed) tried **replacing** the AftSRF with fore→aft cross-attention and failed due to instability. PR #2217 (in progress, nezuko) injects a **single global mean** of fore-surface hidden states into every aft-foil node — a coarse approximation that loses spatial structure.

This experiment tests the middle ground: keep the full AftSRF MLP but enrich each aft-foil node's input with a **per-node cross-attention context vector** drawn from the fore-foil surface hidden states. Each aft node queries the fore surface independently — the model can learn that aft-LE nodes are most affected by the fore-TE wake, while aft-TE nodes are less affected.

Key safety design: the output projection of the cross-attention module is **zero-initialized**, so at epoch 0 the model is identical to the current AftSRF baseline.

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flags

In the `Config` dataclass (near other `aft_foil_srf_*` flags, around line 1120):
```python
aft_srf_fore_crossattn: bool = False    # additive fore->aft cross-attn context in AftSRF
aft_srf_crossattn_dim: int = 64         # attention projection dimension (d_attn)
aft_srf_crossattn_heads: int = 4        # number of attention heads
```

### Step 2: Add `ForeSurfaceCrossAttention` module

Insert as a new class after `AftFoilRefinementHead` (after line 594):

```python
class ForeSurfaceCrossAttention(nn.Module):
    """Per-node cross-attention from aft-foil surface to fore-foil surface hidden states.

    Each aft-foil node queries the full fore-foil surface to get a wake-context vector.
    Zero-init output projection for safe additive integration with AftSRF.
    """

    def __init__(self, n_hidden: int, d_attn: int = 64, n_heads: int = 4):
        super().__init__()
        assert d_attn % n_heads == 0
        self.d_attn = d_attn
        self.n_heads = n_heads
        self.d_head = d_attn // n_heads
        self.q_proj = nn.Linear(n_hidden, d_attn, bias=False)
        self.k_proj = nn.Linear(n_hidden, d_attn, bias=False)
        self.v_proj = nn.Linear(n_hidden, d_attn, bias=False)
        self.out_proj = nn.Linear(d_attn, n_hidden)
        # Zero-init: model starts as pure AftSRF, cross-attn is a no-op at epoch 0
        nn.init.zeros_(self.out_proj.weight)
        nn.init.zeros_(self.out_proj.bias)

    def forward(self, aft_h: torch.Tensor, fore_h: torch.Tensor) -> torch.Tensor:
        """
        Args:
            aft_h:  [A, n_hidden] -- aft-foil surface node hidden states
            fore_h: [F, n_hidden] -- fore-foil surface node hidden states
        Returns:
            [A, n_hidden] -- per-aft-node context vector (zero at init)
        """
        A, F = aft_h.shape[0], fore_h.shape[0]
        H, D = self.n_heads, self.d_head
        Q = self.q_proj(aft_h).view(A, H, D).transpose(0, 1)   # [H, A, D]
        K = self.k_proj(fore_h).view(F, H, D).transpose(0, 1)  # [H, F, D]
        V = self.v_proj(fore_h).view(F, H, D).transpose(0, 1)  # [H, F, D]
        scores = torch.bmm(Q, K.transpose(1, 2)) / (D ** 0.5)  # [H, A, F]
        attn = torch.softmax(scores, dim=-1)                     # [H, A, F]
        ctx = torch.bmm(attn, V).transpose(0, 1).reshape(A, self.d_attn)  # [A, d_attn]
        return self.out_proj(ctx)                                # [A, n_hidden]
```

### Step 3: Instantiate the module

In the model/head initialization block (where `aft_srf_head` is created, search for `AftFoilRefinementHead(`):
```python
fore_crossattn = None
if cfg.aft_srf_fore_crossattn:
    fore_crossattn = ForeSurfaceCrossAttention(
        n_hidden=n_hidden,
        d_attn=cfg.aft_srf_crossattn_dim,
        n_heads=cfg.aft_srf_crossattn_heads,
    ).to(device)
```

Also set up an EMA copy alongside the existing EMA heads:
```python
eval_fore_crossattn = copy.deepcopy(fore_crossattn) if fore_crossattn is not None else None
```

And include it in the optimizer param groups (same group as `aft_srf_head`).

### Step 4: Apply in training loop (critical section)

Find the non-context AftSRF path in the training loop (around line 1886):
```python
elif aft_srf_head is not None and model.training and _aft_foil_mask is not None:
    aft_idx = _aft_foil_mask.nonzero(as_tuple=False)  # [A, 2]
    if aft_idx.numel() > 0:
        aft_hidden = hidden[aft_idx[:, 0], aft_idx[:, 1]]  # [A, n_hidden]
        aft_pred = pred[aft_idx[:, 0], aft_idx[:, 1]]      # [A, 3]
```

Insert cross-attention between the hidden extraction and the AftSRF call:
```python
elif aft_srf_head is not None and model.training and _aft_foil_mask is not None:
    aft_idx = _aft_foil_mask.nonzero(as_tuple=False)  # [A, 2]
    if aft_idx.numel() > 0:
        aft_hidden = hidden[aft_idx[:, 0], aft_idx[:, 1]]  # [A, n_hidden]
        aft_pred = pred[aft_idx[:, 0], aft_idx[:, 1]]      # [A, 3]

        # Additive fore->aft cross-attention (per-sample, since attention is sample-local)
        if fore_crossattn is not None:
            _fore_surf_mask = is_surface & (_raw_saf_norm <= 0.005)  # fore-foil surface
            aft_hidden_enriched = aft_hidden.clone()
            for b in range(hidden.shape[0]):
                aft_b_mask = aft_idx[:, 0] == b
                fore_b_nodes = _fore_surf_mask[b].nonzero(as_tuple=True)[0]
                if aft_b_mask.sum() == 0 or fore_b_nodes.numel() == 0:
                    continue
                aft_h_b = aft_hidden[aft_b_mask]          # [A_b, n_hidden]
                fore_h_b = hidden[b, fore_b_nodes]         # [F_b, n_hidden]
                with torch.amp.autocast("cuda", dtype=torch.bfloat16):
                    ctx_b = fore_crossattn(aft_h_b, fore_h_b).float()
                aft_hidden_enriched[aft_b_mask] = aft_h_b + ctx_b
            aft_hidden = aft_hidden_enriched

        _aft_cond = None
        if cfg.aft_foil_srf_film:
            _aft_cond = _raw_gap_stagger[aft_idx[:, 0]]
        with torch.amp.autocast("cuda", dtype=torch.bfloat16):
            aft_correction = aft_srf_head(aft_hidden, aft_pred, _aft_cond).float()
        pred = pred.clone()
        pred[aft_idx[:, 0], aft_idx[:, 1]] += aft_correction
```

Note: `_raw_saf_norm` is already computed at line 1725 as `x[:, :, 2:4].norm(dim=-1)` (before normalization). Re-use it here.

### Step 5: Mirror in eval loop

Find the eval AftSRF path (search for `eval_aft_srf_head is not None`). Apply the same cross-attention enrichment using `eval_fore_crossattn` (the EMA copy). Mirror the per-sample loop exactly.

### Step 6: EMA update

In the EMA update block (search for `ema_update` or `eval_aft_srf_head` parameter copying), add:
```python
if eval_fore_crossattn is not None and fore_crossattn is not None:
    ema_update(eval_fore_crossattn, fore_crossattn, cfg.ema_decay)
```

### Step 7: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/fore-aft-crossattn-s42" \
  --wandb_group "round16/fore-aft-crossattn-additive" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame \
  --aft_srf_fore_crossattn --aft_srf_crossattn_dim 64 --aft_srf_crossattn_heads 4

# Seed 73
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/fore-aft-crossattn-s73" \
  --wandb_group "round16/fore-aft-crossattn-additive" \
  --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame \
  --aft_srf_fore_crossattn --aft_srf_crossattn_dim 64 --aft_srf_crossattn_heads 4
```

### Step 8: Report results

Add a **Results** comment with:
- Table: p_in, p_oodc, p_tan, p_re per seed and 2-seed average
- W&B run IDs
- Notes on convergence, VRAM overhead, whether cross-attention weights appear to have learned non-trivial patterns

## Baseline

Current best (PR #2207, TE Coordinate Frame, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **12.490** | < 12.49 |
| p_oodc | **7.618**  | < 7.62  |
| **p_tan** | **28.521** | **< 28.52** |
| p_re   | **6.411**  | < 6.41  |

W&B runs: `obn1wfja` (seed 42, p_tan=28.641), `52irfwwg` (seed 73, p_tan=28.400)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-te-coord" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Bug Fix: torch.compile crash with ForeSurfaceCrossAttention

First run (v1) crashed at epoch 20/21 on both seeds with `cudaErrorIllegalAddress`. Root cause: `torch.compile` was applied to the `ForeSurfaceCrossAttention` module, but the per-sample loop inside the training/eval path feeds variable-sized tensors (different fore/aft node counts per sample), causing compiled CUDA kernels to access out-of-bounds memory.

**Fix:** Removed `torch.compile` from `fore_crossattn` module (the module is small — 49K params — so compile overhead would dominate anyway). Relaunched both seeds as v2.

Timestamp: 2026-04-06 ~20:30 UTC

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Key Metrics

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed avg | Delta |
|--------|----------|---------|---------|------------|-------|
| p_in   | 12.490   | 13.3    | 12.4    | **12.850** | +0.360 |
| p_oodc | 7.618    | 7.1     | 7.9     | **7.500**  | -0.118 |
| **p_tan** | **28.521** | 28.6 | 28.0 | **28.300** | **-0.221** |
| p_re   | 6.411    | 6.4     | 6.5     | **6.450**  | +0.039 |

### Surface MAE

| Split | Seed 42 (Ux, Uy, p) | Seed 73 (Ux, Uy, p) |
|-------|---------------------|---------------------|
| val_in_dist | 3.22, 0.82, 13.3 | 2.86, 0.88, 12.4 |
| val_tandem_transfer | —, —, 28.6 | —, —, 28.0 |
| val_ood_cond | —, —, 7.1 | —, —, 7.9 |
| val_ood_re | —, —, 6.4 | —, —, 6.5 |

### Experiment Details

- **Seed 42**: W&B run `2y7ai6nt`, 146 epochs, peak VRAM 47.4 GB
- **Seed 73**: W&B run `8953fle4`, 146 epochs, peak VRAM 47.6 GB
- **ForeSurfaceCrossAttention params**: 49,344 (d_attn=64, 4 heads)
- **Total overhead**: ~1 GB additional VRAM vs baseline (~46 GB)

### Commands used

```bash
# Seed 42 (GPU 2)
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=2 python train.py \
  --agent alphonse --wandb_name "alphonse/fore-aft-crossattn-s42-v5" \
  --wandb_group "round16/fore-aft-crossattn-additive" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame \
  --aft_srf_fore_crossattn --aft_srf_crossattn_dim 64 --aft_srf_crossattn_heads 4

# Seed 73 (GPU 4) — same flags with seed 73
```

### What happened

**Mixed result — small improvement on p_tan (-0.221) and p_oodc (-0.118), but regression on p_in (+0.360) and flat on p_re.**

The fore→aft cross-attention adds per-node spatially-resolved wake context to each aft-foil node before the AftSRF correction MLP. The hypothesis that aft-foil pressure benefits from spatial awareness of the fore-foil surface is partially supported — the tandem-specific metrics (p_tan, p_oodc) improve while the in-distribution metric (p_in) regresses.

**Important implementation caveats:**

1. **torch.compile crash**: The cross-attention's per-sample loop with variable-length fore/aft surface node counts triggers intermittent `cudaErrorIllegalAddress` crashes when interacting with compiled modules. Required workarounds:
   - `fore_crossattn` module is NOT compiled (variable tensor shapes)
   - `aft_srf_head` is NOT compiled when cross-attention is active (compiled head + enriched tensors = crash)
   - Hidden states are `.detach().clone()`d before cross-attention to break graph references from compiled base model
   
2. **Gradient flow limitation**: Due to the detach/clone workaround, gradients do NOT flow from the cross-attention back through the base model's hidden states. The cross-attention module's parameters still learn (through the output loss path), but the base model cannot adjust its hidden representations to better serve the cross-attention. This likely limits the approach's effectiveness.

3. **Convergence**: Both seeds converged smoothly to 146 epochs with no issues after the workarounds. Training time ~74s/epoch, comparable to baseline.

### Suggested follow-ups

1. **Fix the gradient flow**: The detach/clone is a workaround. A proper fix might use `torch._dynamo.config.suppress_errors = True` or restructure to avoid torch.compile interactions. With proper gradient flow, the model could learn to produce hidden states that are more informative for the cross-attention.

2. **Try a simpler approach**: A mean-pooled fore-surface context (like #2217/nezuko's approach but for the aft SRF specifically) might capture most of the benefit without the torch.compile complications.

3. **Larger d_attn**: The current 64-dim attention might be too small. A 128-dim or 256-dim attention with 8 heads could capture richer spatial patterns.

4. **Positional encoding**: Adding relative position features (distance from fore-TE to each aft node) to the cross-attention Q/K could help the model learn more physically meaningful attention patterns.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2219 because Mixed/marginal results that don't beat the current baseline. Against the current baseline (PR #2213, wake deficit): p_in +7.3%, p_re +2.4% worse. Only p_oodc improved (-1.9%). p_tan at 28.300 vs baseline 28.341 is essentially flat (well within seed noise). This experiment also ran on the outdated TE-only baseline (missing --wake_deficit_feature).

The torch.compile crash and required workarounds (no compile on cross-attn + SRF head, detach/clone blocking gradient flow) are structural issues that limit this approach. Without gradient flow from the cross-attention back to the backbone, the model can't adapt its hidden representations to be more informative for the cross-attention — this caps the potential upside.

**Fore→aft information flow direction now tested 3 ways:**
- #2202: replacement cross-attention → failed (p_tan +2.1%, unstable)
- #2217: mean-pooled skip → failed (3/4 metrics worse)
- #2219: additive cross-attention → marginal (p_oodc improved, rest flat/worse)

All three show consistent p_oodc improvement but nothing else. The p_oodc gain likely comes from adding more learnable parameters near the AftSRF head, not specifically from the cross-attention mechanism. The backbone's attention mechanism already handles fore-aft coupling adequately. This direction is exhausted.

Good work on the bug fix (torch.compile crash) and thorough analysis of the gradient flow limitation..
